### ISSUE TITLE
Nested field support

### ISSUE THREAD

**Issue by jasonbahl** [jakepusateri/graphql-list-fields#2]
If I have a query like this: 

```
{
   somequery{
      field1
      field2
      {
         nestedField
      }
   }
}
```

I only get `field1` and `field2` returned, not the nested fields. . .is that a bug or am I missing something in using it?

**Comment by jakepusateri** [comment]
I don't think the query you posted is valid. If we take this query instead:
```
{
   somequery {
      field1
      field2
      field3 {
         nestedField
      }
   }
}
```
What would you expect this to return? It currently returns ['field1', 'field2']. 

['field1', 'field2', 'field3.nestedField'] might make sense though, but it should always return a flat array.

**Comment by jasonbahl** [comment]
ya, you're correct, my formatting was off.

So, one project I'm working on is a central GraphQL server that communicates to several other GraphQL servers, so we can query a single endpoint and specify in the $args which other systems to get data from, and behind the scenes the resolver will take the query that was entered, use it to query the other GraphQL servers, combine the results and return them. 

So what I'm trying to accomplish is to get the GraphQL query language format of the "subfields" that were entered for any given field.

So, ideally, a query such as: 

```
{
   somequery {
      field1
      field2
      field3 {
         nestedField
      }
   }
}
```

In the resolve function for the `somequery` field, I'm hoping to have: 

```
      field1
      field2
      field3 {
         nestedField
      }
```

and ideally in the query language format so that I can pass it as a query to the distributed systems I'm grabbing data from.

I ended up getting close with the GraphQL.js `print` function, but it gets funky with fragments. 

Any thoughts? I've got a feeling your project isn't exactly what I'll need, but perhaps can get me close-ish?

**Comment by jakepusateri** [comment]
Hmm, this project won't get you the subquery for a field in the graphql format again, especially if you want to preserve query variables or fragments.

**Comment by sudheerj** [comment]
How we need to deal with nested object fields now? It throws as error if the query contains nested object

> column parent.child does not exist

**Comment by jakepusateri** [comment]
Can you provide an example query that generates that error?

**Comment by sudheerj** [comment]
For example below is the query

accounts(offset: 0,limit: 2,order: "account_number DESC",filter:"{'account_number':'4521','accounts.person': {'$in': [ 'jack', 'john' ]}}"){
       account_number,
       account_type_code,
  	   parent_ucm_id,
       account_status,
       streamlined_flag,
       account_name,
       person {
         name
       },
       region
    }

I can see only ibxes as a field but not nested object fields information.

**Comment by TdyP** [comment]
Any plan on supporting nested fields?
My query is quite simple: 
```
{
  users {
    identifier
    login
    friend {
      login
    } 
  }
}
```

I need to know which fields are requested under `friend`

In my imagination, results should be something like that: 
```['identifier', 'login', { friend: ['login'] } ]```

**Comment by jakepusateri** [comment]
Yes, I do have a pending PR that might fix it. I haven't merged it because I can't decide whether or not to include the root node in the returned list. What is your opinion on that?

**Comment by TdyP** [comment]
I don't think the root node should be included in the node. This list is used is basically used inside a resolver, so we know what we're resolving.

I'm more concerned about the use of dot-notation. It involves to do some string parsing afterwards and it has some major limitation, like support of field arguments for example.
My preference would go to returning something like this:
```
{
    name: 'label'
},
{
    name: 'id'
},
{
    name: 'upsell',
    fields: [
        {
            name: 'label'
        },
        {
            name: 'crosssell',
            fields: [
                {
                    name: 'label'
                },
                {
                    name: 'price',
                    arguments: {
                        arg1: 'value'
                    }
                }
            ]
        }
    ]
}
```

But I'm totally aware that would be a major breaking change for your library

**Comment by cdelgadob** [comment]
From my point of view, a boolean argument would be the way to go, indicating whether nested fields would be required or not.

In my case, I have a graphQL server backed by a DynamoDB, and I want each resolver to gather only the requested fields to each table, saving costs and time.

If there's a nested objet, from another table, I will delegate the field filtering to that object's resolver.

This said, the plan is as follows:
Given this definition:

    export const user = new GraphQLObjectType({
      name: "user",
       fields: () => ({
        email: {type: GraphQLID},
        registered_date: {type: GraphQLString},
        person: {type: personModel} // Another graphql object
      })
    });

The user resolver would fetch and pass the requested fields using the `getFieldNames(info)` function from the user table to the underlying database access, and the personModel resolver will be called only if that field is in the query and then, again, it's own resolver will fetch only the fields requested for that object.

@jakepusateri Does this sound ok to you?

Re the nested fields, I'd go for the '.' notation, but only if nested fields are requested.

**Comment by jakepusateri** [comment]
Closed via #3
